src/record_ts.cpp

```diff
@@ -0,0 +1,1446 @@
+
+                        jitc_log(
+                            LogLevel::Debug,
+                            "replay(): Inferring launch size by heuristic, "
+                            "launch_size=%u, ratio=%u",
+                            launch_size, ratio);
+                        launch_size = launch_size * ratio;
+                    } else if (op.input_size % op.size == 0) {
+                        uint32_t fraction = op.input_size / op.size;
+                        jitc_log(
+                            LogLevel::Debug,
+                            "replay(): Inferring launch size by heuristic, "
+                            "launch_size(%u), fraction=%u",
+                            launch_size, fraction);
+                        launch_size = launch_size / fraction;
+                    }
+                }
+                if (launch_size == 0) {
+                    jitc_log(LogLevel::Debug,
+                             "replay(): Could not infer launch "
+                             "size, using recorded size");
+                    launch_size = op.size;
+                }
+
+                // Allocate output variables for kernel launch.
+                // The assumption here is that for every kernel launch, the
+                // inputs are already allocated. Therefore we only allocate
+                // output variables, which have the same size as the kernel.
+                for (uint32_t j = op.dependency_range.first;
+                     j < op.dependency_range.second; ++j) {
+                    ParamInfo info     = this->dependencies[j];
+                    ReplayVariable &rv = replay_variables[info.slot];
+
+                    if (info.type == ParamType::Input) {
+                        uint32_t size = rv.size(info.vtype);
+                        jitc_log(LogLevel::Info,
+                                 " -> param s%u is_pointer=%u size=%u",
+                                 info.slot, info.pointer_access, size);
+                    } else {
+                        jitc_log(LogLevel::Info, " <- param s%u is_pointer=%u",
+                                 info.slot, info.pointer_access);
+                    }
+
+                    if (info.type == ParamType::Output) {
+                        rv.alloc(backend, launch_size, info.vtype);
+                    }
+                    jitc_log(LogLevel::Info, "    data=%p", rv.data);
+                    jitc_assert(
+                        rv.data != nullptr || dry_run,
+                        "replay(): Encountered nullptr in kernel parameters.");
+                    kernel_params.push_back(rv.data);
+                }
+
+                // Change kernel size in `kernel_params`
+                if (backend == JitBackend::CUDA) {
+                    uintptr_t size = 0;
+                    std::memcpy(&size, &launch_size, sizeof(uint32_t));
+                    kernel_params[0] = (void *) size;
```

Comment:
I'm not 100% sure I understand this. Is it essentially writing `launch_size` to `kernel_params[0]`, but masking out bits 32...63?
Wouldn't it be better to throw an exception if `launch_size >= 2**32` and directly write `kernel_params[0] = launch_size;`?

Reply:
I just copied what is in `eval.cpp`:
```cpp
    if (backend == JitBackend::CUDA) {
        uintptr_t size = 0;
        memcpy(&size, &group.size, sizeof(uint32_t));
        kernel_params.push_back((void *) size);

        // The first 3 variables are reserved on the CUDA backend
        n_regs = 4;
    } else {...
```
As far as I understand, the `memcpy` ensures that we copy to the first 4 bits of the first void pointer in `kernel_params` (endianess?).